### PR TITLE
ENH:linalg: Add LAPACK wrappers pptrf/pptrs/ppsv/pptri/ppcon

### LINKED ISSUE #12431: Request for Python wrapper for LAPACK's ?pptrf (Cholesky factorization in packed format)
**Is your feature request related to a problem? Please describe.**

scipy.linalg.lapack already has wrappers for ?tpsv (triangular solve in packed format).  ?pptrf (Cholesky factorization in packed format) seems to be a common enough function that can be used to generate the matrix passed to ?tpsv, but it is not wrapped at the Python level.

**Describe the solution you'd like**

Add a wrapper for ?pptrf.

**Describe alternatives you've considered**

Manually converting between unpacked and packed formats, but this is not exactly fast.

### PR DESCRIPTION
**Pull request by ilayn**
As title mentions.

Closes #12431

### CODE REVIEW

**Comment by mdhaber** [comment]
Looks like a tolerance issue.
```
C:\hostedtoolcache\windows\Python\3.6.8\x64\lib\site-packages\scipy\linalg\tests\test_lapack.py:2945: in test_pptrs_pptrf_ppsv
    assert_allclose(ul, aul, rtol=np.finfo(dtype).eps*50)
E   AssertionError: 
E   Not equal to tolerance rtol=1.11022e-14, atol=0
E   
E   Mismatched elements: 1 / 55 (1.82%)
E   Max absolute difference: 1.11022302e-16
E   Max relative difference: 8.776365e-14
E    x: array([ 2.320138e+00,  4.223568e-01,  2.413215e+00,  3.459337e-01,
E           4.776830e-01,  2.191555e+00,  7.130982e-01,  3.512715e-01,
E           3.451844e-01,  2.129902e+00,  3.619548e-01,  1.664434e-01,...
E    y: array([ 2.320138e+00,  4.223568e-01,  2.413215e+00,  3.459337e-01,
E           4.776830e-01,  2.191555e+00,  7.130982e-01,  3.512715e-01,
E           3.451844e-01,  2.129902e+00,  3.619548e-01,  1.664434e-01,...
```

File: scipy/linalg/tests/test_lapack.py [hunk 1]
```diff
@@ -2923,3 +2924,29 @@ def test_ptsvx_NAG(d, e, b, x):
...
+
+
+@pytest.mark.parametrize('dtype', DTYPES)
+def test_pptrs_pptrf_ppsv(dtype):
+    seed(1234)
+    # Manual conversion to/from packed format is feasible here.
+    n, nrhs = 10, 4
+    a = generate_random_dtype_array([n, n], dtype=dtype)
+    b = generate_random_dtype_array([n, nrhs], dtype=dtype)
+    a = a.conj().T + a + np.eye(n, dtype=dtype) * dtype(5.)
```
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py:2936]:
```suggestion
    a = a.conj().T + a + np.eye(n, dtype=dtype) * dtype(5.)
```
I suppose this happens to make it diagonally dominant, but that depends on the RNG gods, right? I know you're setting the seed so if it works, it works; just checking.

File: scipy/linalg/tests/test_lapack.py [hunk 2]
```diff
@@ -2923,3 +2924,29 @@ def test_ptsvx_NAG(d, e, b, x):
...
+    inds = ([x for y in range(1, n+1) for x in range(y)],
+            [y-1 for y in range(1, n+1) for x in range(y)])
+    ap = a[inds]
+    ppsv, pptrf, pptrs = get_lapack_funcs(('ppsv', 'pptrf', 'pptrs'),
+                                          dtype=dtype)
+
+    ul, info = pptrf(n, ap)
+    aul = cholesky(a)[inds]
+    assert_allclose(ul, aul, rtol=np.finfo(dtype).eps*50)
```
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py:2945]: Tests for correct factorization.  👍

File: scipy/linalg/tests/test_lapack.py [hunk 3]
```diff
@@ -2923,3 +2924,29 @@ def test_ptsvx_NAG(d, e, b, x):
...
+
+    x, info = pptrs(n, ul, b)
+    bx = solve(a, b)
+    assert_allclose(x, bx, rtol=np.finfo(dtype).eps*50)
```
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py:2949]: Tests for correct solution 👍

File: scipy/linalg/tests/test_lapack.py [hunk 4]
```diff
@@ -2923,3 +2924,29 @@ def test_ptsvx_NAG(d, e, b, x):
...
+
+    xv, info = ppsv(n, ap, b)
+    assert_allclose(xv, bx, rtol=np.finfo(dtype).eps*50)
```
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py:2952]: Tests for correct solution 👍

File: scipy/linalg/tests/test_lapack.py [hunk 5]
```diff
@@ -2923,3 +2924,29 @@ def test_ptsvx_NAG(d, e, b, x):
...
+            [y-1 for y in range(1, n+1) for x in range(y)])
+    ap = a[inds]
+    ppsv, pptrf, pptrs = get_lapack_funcs(('ppsv', 'pptrf', 'pptrs'),
+                                          dtype=dtype)
+
+    ul, info = pptrf(n, ap)
+    aul = cholesky(a)[inds]
+    assert_allclose(ul, aul, rtol=np.finfo(dtype).eps*50)
+
+    x, info = pptrs(n, ul, b)
```
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py:2947]: Shouldn't the routine be able to figure out `n`?

**Review by mdhaber — COMMENTED** [review]
Should test `UPLO`, and can't the wrapper figure out `n`?

File: scipy/linalg/tests/test_lapack.py [hunk 6]
```diff
@@ -2923,3 +2924,29 @@ def test_ptsvx_NAG(d, e, b, x):
     df, ef, x_ptsvx, rcond, ferr, berr, info = ptsvx(d, e, b)
     # determine ptsvx's solution and x are the same.
     assert_array_almost_equal(x, x_ptsvx)
+
+
+@pytest.mark.parametrize('dtype', DTYPES)
```
[the remark this answers is not in the document]
Comment (ilayn) [reply scipy/linalg/tests/test_lapack.py:2929]: Bah, I forgot lower keyword. Will add.

File: scipy/linalg/tests/test_lapack.py (hunk 1 above)
Comment (ilayn) [reply scipy/linalg/tests/test_lapack.py:2936]: For uniformly sampled random symmetric matrices of 10x10, all eigs except the largest ones are centered around zero following Wigner's semicircle law. So 5 is ~2x the minimum eigenvalue that is expected.

File: scipy/linalg/tests/test_lapack.py (hunk 5 above)
Comment (ilayn) [reply scipy/linalg/tests/test_lapack.py:2947]: See #12333 for a similar discussion

**Comment by sturlamolden** [comment]
Wrappers should have the threadsafe declaration to release the GIL. We assume LAPACK is threadsafe.

**Comment by sturlamolden** [comment]
Hm, ```threadsafe``` seems to be applied rather randomly to BLAS and LAPACK, but it should be applied to all of LAPACK and BLAS.

**Comment by ilayn** [comment]
I have no idea about what it does but I'll take your word for it. Adding them now

**Comment by sturlamolden** [comment]
> I have no idea about what it does but I'll take your word for it. Adding them now

It makes the wrapper release the GIL, which we should only do if the Fortran code is reentrant.

Fortran functions and subroutines are normally not re-entrant unless declared ```RECURSIVE``` or compiled with ```-frecursive``` (or something with similar effect). This is because the Fortran compiler is otherwise free to give local arrays static storage instead of allocating them on the stack or the heap (aka implict ```SAVE```). A notable compiler which makes use of this evil is gfortran. We have (in other places) assumed that BLAS and LAPACK are compiled with the correct flags to be reentrant. At least this is the case with OpenBLAS and MKL. 

BLAS is reentrant with ATLAS as well, but those who use it might have to compile LAPACK with ```-frecursive```.

**Comment by ilayn** [comment]
Added `threadsafe`s, corrected the IPL_64 integer stuff, added pptri/ppcon (because why not while we are at it).

File: scipy/linalg/tests/test_lapack.py (hunk 1 above)
Comment (sturlamolden) [reply scipy/linalg/tests/test_lapack.py:2939]: Cholesky needs the matrix to be symmetric as well as positive definite, so this should be
```a = a.conj().T @ a + np.eye(n, dtype=dtype) * dtype(5.)```

**Review by sturlamolden — CHANGES_REQUESTED** [review]

File: scipy/linalg/tests/test_lapack.py (hunk 1 above)
Comment (ilayn) [reply scipy/linalg/tests/test_lapack.py:2939]: Probably I am blinded but isn't `a + a*` already hermitian and adding `5*I` makes it pos def?

File: scipy/linalg/tests/test_lapack.py [hunk 7]
```diff
@@ -2923,3 +2924,50 @@ def test_ptsvx_NAG(d, e, b, x):
...
+    if lower:
+        inds = ([x for y in range(n) for x in range(y, n)],
+                [y for y in range(n) for x in range(y, n)])
+    else:
+        inds = ([x for y in range(1, n+1) for x in range(y)],
+                [y-1 for y in range(1, n+1) for x in range(y)])
+    ap = a[inds]
+    ppsv, pptrf, pptrs, pptri, ppcon = get_lapack_funcs(('ppsv', 'pptrf',
+                                                         'pptrs', 'pptri',
+                                                         'ppcon'), dtype=dtype)
```
Comment (pv) [inline scipy/linalg/tests/test_lapack.py:2949]: If you add `ilp64="preferred"` here, also that path gets tested on the CI. (Or, parametrize the test; `scipy.linalg.blas.HAS_ILP64` is the relevant enable flag.)

File: scipy/linalg/tests/test_lapack.py (hunk 1 above)
Comment (mdhaber) [reply scipy/linalg/tests/test_lapack.py:2939]: Yes.

File: scipy/linalg/tests/test_lapack.py (hunk 5 above)
Comment (mdhaber) [reply scipy/linalg/tests/test_lapack.py]: By the time I got there, I forgot it was in a packed format.

File: scipy/linalg/tests/test_lapack.py (hunk 7 above)
Comment (ilayn) [reply scipy/linalg/tests/test_lapack.py:2949]: Ah sure. This will take some time for me to get used to this.

**Comment by ilayn** [comment]
If there are no more comments I'd like to merge this soon.

**Comment by ilayn** [comment]
OK here we go then.

**Comment by mdhaber** [comment]
Thanks @ilayn!

**Pull request merged**